src/blueprints/tv2/tv2-on-timeline-generate-calculator.ts

```diff
@@ -0,0 +1,288 @@
+      if (!mediaPlayerSessionInUse) {
+        return
+      }
+      this.updateTimelineObjectWithMediaPlayer(blueprintTimelineObject, mediaPlayerSessionInUse.mediaPlayer)
+    })
+  }
+
+  private assignMediaPlayersForGroup(group: TimelineObjectGroup, mediaPlayerSessionsInUse: Tv2MediaPlayerSession[], availableMediaPlayers: Tv2MediaPlayer[]): void {
+    const timelineObjects: TimelineObject[] = this.flatMapTimelineObjectChildren(group)
+    timelineObjects.forEach(timelineObject => {
+      const blueprintTimelineObject: Tv2BlueprintTimelineObject = timelineObject as Tv2BlueprintTimelineObject
+      if (!blueprintTimelineObject.metaData || !blueprintTimelineObject.metaData.mediaPlayerSession) {
+        return
+      }
+      const mediaPlayerSessionInUse: Tv2MediaPlayerSession | undefined = mediaPlayerSessionsInUse.find(session => session.sessionId === blueprintTimelineObject.metaData?.mediaPlayerSession)
+      if (mediaPlayerSessionInUse) {
+        this.updateTimelineObjectWithMediaPlayer(blueprintTimelineObject, mediaPlayerSessionInUse.mediaPlayer)
+        return
+      }
+      const mediaPlayer: Tv2MediaPlayer | undefined = availableMediaPlayers.pop()
+      if (!mediaPlayer) {
+        return
+      }
+      mediaPlayerSessionsInUse.push({
+        mediaPlayer,
+        sessionId: blueprintTimelineObject.metaData.mediaPlayerSession
+      })
+      this.updateTimelineObjectWithMediaPlayer(blueprintTimelineObject, mediaPlayer)
+    })
+  }
+
+  private findPreviousAssignedMediaPlayerSessionsStillInUseForGroup(assignedMediaPlayerSessions: Tv2MediaPlayerSession[], group: TimelineObjectGroup): Tv2MediaPlayerSession[] {
+    return assignedMediaPlayerSessions.filter(mediaPlayerSession => {
+      return group.children.some(child => this.doesTimelineObjectHaveMediaPlayerSessionWithId(child, mediaPlayerSession.sessionId))
+    })
+  }
+
+  private doesTimelineObjectHaveMediaPlayerSessionWithId(timelineObject: TimelineObject, sessionId: string): boolean {
+    let doesChildrenHaveMediaPlayerSession: boolean = false
+    if (timelineObject.children) {
+      doesChildrenHaveMediaPlayerSession = timelineObject.children.some((child: TimelineObject) => this.doesTimelineObjectHaveMediaPlayerSessionWithId(child, sessionId))
+    }
+    const blueprintTimelineObject: Tv2BlueprintTimelineObject = timelineObject as Tv2BlueprintTimelineObject
+    const hasMediaPlayerSession: boolean = blueprintTimelineObject.metaData?.mediaPlayerSession === sessionId
+
+    return hasMediaPlayerSession || doesChildrenHaveMediaPlayerSession
+  }
+
+  private flatMapTimelineObjectChildren(timelineObject: TimelineObject): TimelineObject[] {
+    if (timelineObject.children) {
+      return timelineObject.children.flatMap((child: TimelineObject) => this.flatMapTimelineObjectChildren(child))
+    }
+    return [timelineObject]
+  }
```

Comment:
Yes, but only the children of the very last "node", since they are the actual TimelineObjects.

A TimelineGroup consist of "the main object" e.g. `ActiveGroup`.
This main object has a "control" TimelineObject that it is used to decide when the group is on the Timeline when resolving.
For each Piece of the Part, the main object another TimelineObject which is the "PieceGroup". Each of these "PieceGroups" has an array of TimelineObjects, which are the actual TimelineObjects that we want to send to TSR.

This "hierarchy" is how Core also does this. I'm going to write a Confluence page about this once I'm done implementing building the Timeline.

If you have a suggestion for another way to fetch the actual TimelineObjects into a flat array I'm all ears :)

Reply:
Understandably. It's been renamed.